src/lib.rs

```diff
@@ -236,129 +181,140 @@ impl<'a> Interpreter<'a> {
     ///
     /// Interpreters are relatively large, so avoid too many calls to this
     /// function.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use brainfuck::Interpreter;
-    ///
-    /// let mut reader = &[][..];
-    /// let mut writer = Vec::<u8>::new();
-    /// Interpreter::new("+.", &mut reader, &mut writer).unwrap().run();
-    /// ```
-    pub fn new<C: Into<String>, R: Read, W: Write>(code: C, input: &'a mut R, output: &'a mut W) -> Result<Interpreter<'a>, Error> {
-        // TODO: Compress and cache the code, removing everything but code.
-        //       This will allow running to avoid the overhead of finding
-        //       instructions and brace matching.
-        Ok(Interpreter {
-            code: code.into(),
+    pub fn new<R: Read, W: Write>(input: &'a mut R, output: &'a mut W) -> Interpreter<'a> {
+        Interpreter {
+            program: None,
             reader: input,
             writer: output,
             tape: [0; 30000],
             ptr: 0,
             pc: 0,
-        })
+        }
     }
 
-    /// Create a new interpreter from a file.
-    ///
-    /// Loads the given file from the path, and creates a new
-    /// interpreter. The code of the interpreter is the contents
-    /// of the file. Every cell of the tape is initialized to 0.
-    /// Both the pointer and the program counter are set to 0. If
-    /// the file fails to load then this function returns an error.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use brainfuck::Interpreter;
-    ///
-    /// let mut reader = &[][..];
-    /// let mut writer = Vec::<u8>::new();
-    /// let interp = Interpreter::from_file("fixtures/hello.b", &mut reader, &mut writer);
-    /// ```
-    pub fn from_file<P: AsRef<Path>, R: Read, W: Write>(path: P, input: &'a mut R, output: &'a mut W) -> Result<Interpreter<'a>, Error> {
-        let mut file = try!(File::open(path));
-        let mut code = String::new();
-        try!(file.read_to_string(&mut code));
-        Interpreter::new(code, input, output)
+    /// Load a program for the interpreter to run.
+    pub fn load(&mut self, program: Program) -> &mut Self {
```

Comment:
I think a bunch of thought can be given to a nice user interface for this. I was thinking maybe something like a top level `brainfuck::eval` or something like that. It would create a program and load it into the interpreter with stdin and stdout bound.